cachi2/core/package_managers/bundler/parser.py

```diff
@@ -80,6 +80,44 @@ def download_to(self, deps_dir: RootedPath) -> None:
         download_binary_file(self.remote_location, fs_location)
 
 
+class GemPlatformSpecificDependency(GemDependency):
+    """
+    Represents a gem dependency built for a specific platform.
+
+    Attributes:
+        platform:     Platform for which the dependency was built.
+    """
+
+    platform: str
+
+    @property
+    def remote_location(self) -> str:
+        """Return remote location to download this gem from."""
+        return f"{self.source}/downloads/{self.name}-{self.version}-{self.platform}.gem"
+
+    def download_to(self, deps_dir: RootedPath) -> None:
+        """Download represented gem to specified file system location."""
+        fs_location = deps_dir.join_within_root(
+            Path(f"{self.name}-{self.version}-{self.platform}.gem")
+        )
+        log.info(
+            "Downloading platform-specific gem %s-%s-%s", self.name, self.version, self.platform
+        )
+        try:
+            download_binary_file(self.remote_location, fs_location)
+        except FetchError:
+            # A combination of Ruby v.3.0.7 and some Bundler dependencies results in
+            # -gnu suffix being dropped from some platforms. This was observed on
+            # sqlite3-aarch-linux-gnu. We cannot control user's Ruby version,
+            # but we could try and guess correct path. If this fails then we should
+            # assume that package is broken.
+            self.platform = self.platform + "-gnu"
+            fs_location = deps_dir.join_within_root(
+                Path(f"{self.name}-{self.version}-{self.platform}.gem")
+            )
+            download_binary_file(self.remote_location, fs_location)
```

Comment:
Based on my limited research above (missing `musl` target in the lockfile) I am not convinced that we want to assume this problem exists only for `-gnu`. The reporter of https://github.com/rubygems/rubygems/issues/7432 mentioned that he could work around the problem by explicitly doing `bundle lock --add-platform xyz-musl` and then it produced the expected results. At best I'd suggest keeping a neutral attitude here, not trying to recover from an error that likely doesn't originate at cachi2, and simply re-raise  `FetchError` (or change the type) providing a solution and in that solution mention that the lockfile should be inspected and the actual URL checked for existence.